lib/scholar/linear_regression.ex

```diff
@@ -0,0 +1,42 @@
+defmodule Scholar.LinearRegression do
+
+  defstruct :coefficients
+
+  @doc """
+  Fits a linear regression model to the given dataset.
+  """
+  def fit(x, y) do
+    {coeff, intercept} = lstsq(x, y)
+  end
+
+  defnp lstsq(x, y) do
+    {u, s, vt} = Nx.LinAlg.svd(x)
+
+    uty =
+      u
+      |> Nx.transpose()
+      |> Nx.dot(y)
+
+    s_inv = Nx.new_axis(1 / s, -1)
+
+    x =
+      vt
+      |> Nx.transpose()
+      |> Nx.dot(Nx.multiply(s_inv, uty))
```

Comment:
The overall linalg seems right! Although I had never seen this formulation, going through `At . A = At.b` with `A = U.S.Vt` yielded the same result.

I got to this other implementation which has some parts which are a little bit more direct (without needing to use Nx.new_axis). Perhaps there's something there you like?

```elixir
iex(9)> a                                                                                           
#Nx.Tensor<
  s64[3][3] 
  Torchx.Backend(cpu)
  [
    [1, 1, 1],
    [0, 2, 1],
    [0, 0, 3]
  ]
>
iex(10)> b
#Nx.Tensor<
  s64[3]
  Torchx.Backend(cpu)
  [3, 5, 9]
>
iex(11)> Nx.LinAlg.solve(a, b)
#Nx.Tensor<
  f32[3]
  Torchx.Backend(cpu)
  [-1.0, 1.0, 3.0]
>
iex(12)> {u, s, vt} = Nx.LinAlg.svd(a)                                                               
{#Nx.Tensor<
   f32[3][3]
   Torchx.Backend(cpu)
   [
     [0.4022642970085144, 0.3421006500720978, 0.8492057919502258],
     [0.4796755611896515, 0.711304783821106, -0.5137672424316406],
     [0.7798042893409729, -0.6140136122703552, -0.12203514575958252]
   ]
 >,
 #Nx.Tensor<
   f32[3]
   Torchx.Backend(cpu)
   [3.520359516143799, 1.9629522562026978, 0.868269681930542]
 >,
 #Nx.Tensor<
   f32[3][3]
   Torchx.Backend(cpu)
   [
     [0.11426796764135361, 0.3867832124233246, 0.915063738822937],
     [0.1742786318063736, 0.8990083336830139, -0.4017597734928131],
     [0.9780439138412476, -0.20538432896137238, -0.03531980887055397]
   ]
 >}
iex(13)> vt |> Nx.transpose() |> Nx.multiply(Nx.divide(1, s)) |> Nx.dot(Nx.transpose(u)) |> Nx.dot(b)
#Nx.Tensor<
  f32[3]
  Torchx.Backend(cpu)
  [-0.9999991655349731, 0.9999992847442627, 2.999999761581421]
>
```

PS: In retrospect, the final formulation is just the same as using x = inv(A).b where inv(A) is approximated through SVD :thinking:

Reply:
@cigrainger w.r.t checking the data, we don't really need to check the contents of the matrices. We could support both tall and wide a matrices (which neither code seems to support right now). Brace for incoming linear algebra :sweat_smile: 

Let's suppose we have A{n, m} and b{n, k}, which yields x{m, k}.

For the A matrix with shape {n, m}, Nx.LinAlg.svd returns:
U: {n, n}
S: {m}
Vt: {m, m}

We would need to transform the S vector into an {n, m} diagonal matrix before applying the `iex(13)` line above, and it would need to change its Nx.multiply into Nx.dot with the transposed S matrix. Something like:

`vt |> Nx.transpose() |> Nx.dot(Nx.divide(1, s_as_diagonal(s, n, m))) |> Nx.dot(Nx.transpose(u)) |> Nx.dot(b)`

with:

```elixir
def s_as_diagonal(t, n, m) do
  # Note that this returns the transposed shape to be used directly where we need it
  [t, Nx.broadcast(0, {abs(m - n)})]
  |> Nx.concatenate()
  |> Nx.make_diagonal()
  |> Nx.slice([0, 0], [m, n])
end
```

Reply:
PS: Some NaNs/Infs can appear due to `Nx.divide(1, s)` if s contains 0-valued singular values. This doesn't crash the code, but will return an unusable result. I don't think there's too much we can do though

Reply:
Makes sense! I think it's a good idea to warn in that case?

Reply:
Do you mean we would print something or raise? 
Perhaps we could return a flag which indicates if the result is numerically valid together with the data of interest?

We would need to add something like `Nx.has_nan` and `Nx.has_inf` or even `Nx.numerically_valid` though, to make this check work.

Reply:
Actually, nevermind, we just need to `Nx.any(Nx.less_equal(s, eps))` and operate from there

Reply:
Print something I think -- e.g. [`IO.warn/1`](https://hexdocs.pm/elixir/1.12/IO.html#warn/1).

Reply:
I was using the least squares implementation that JAX has which is somewhat what I've implemented here. As far as ensuring s is safe, we'll just need a fuzz value:

```elixir
mask = ...some check that s is safe...
safe_s = Nx.select(mask, s, eps)
```

Where eps is some small value